File: Wilson_grafo.py
import random
import threading


qtde = 0
menorValor = 0
caminhoMaisRapido = ""
matriz = 0
threads = []
primeiraVezPassando = True


def acharCaminho(
    pInicial,
    pFinal,
    caminho,
    valorAresta=0,
):
    global matriz
    contador = 0
    if pInicial == pFinal:
        global qtde, menorValor, caminhoMaisRapido, primeiraVezPassando
        caminho = [x + 1 for x in caminho]
        caminho = " -> ".join(map(str, caminho))
        if primeiraVezPassando:
            primeiraVezPassando = False
            menorValor = valorAresta
        elif menorValor > valorAresta:
            menorValor = valorAresta
            caminhoMaisRapido = caminho

        qtde += 1
        print(f"{caminho}: {valorAresta}\n")
        return 0
    for _ in matriz[pInicial]:
        valor = matriz[pInicial][contador]

        if contador not in caminho and valor > 0:

            global threads
            thread = threading.Thread(
                target=acharCaminho,
                args=(
                    contador,
                    pFinal,
                    caminho.copy() + [contador],
                    valorAresta + valor,
                ),
            )
            threads.append(thread)
            thread.start()

        contador += 1


def iniciarMatriz():
    numRange = int(input("range: "))
    lista = [[0 for _ in range(numRange)] for _ in range(numRange)]
    for i in range(numRange):
        for j in range(numRange):
            if i == j:
                lista[i][j] = 0
            else:
                lista[i][j] = 0
    return lista


def main():
    global matriz
    while True:
        escolha = int(
            input(
                "\n1: Iniciar a Matriz\n2: Mostrar a Matriz\n3: Editar a Matriz\n4: Números Aleatórios\n5: Achar caminho\n6: Excluir a Matriz\n"
            )
        )

        if escolha == 1:
            matriz = iniciarMatriz()
        elif escolha == 2 and isinstance(matriz, list):
            for i in matriz:
                print(i)
        elif escolha == 3 and isinstance(matriz, list):
            orientadoBool = input("É orientado ou não? (Y/N)").upper() == "Y"
            arestas = int(input("Quantas arestas?\n"))
            for i in range(arestas):
                # Escrever no formato p1,p2,valor
                x, y, valor = map(int, input("Índices e valor: ").split(","))
                if y == x:
                    print("Os índice não podem ser iguais")
                else:
                    matriz[x - 1][y - 1] = valor
                    if not orientadoBool:
                        matriz[y - 1][x - 1] = valor

        elif escolha == 4 and isinstance(matriz, list):
            lenght = len(matriz)
            for i in range(lenght):
                for j in range(lenght):
                    if not i == j:
                        matriz[j][i] = random.randrange(0, 10)
        elif escolha == 5 and isinstance(matriz, list):
            pInicial = int(input("Ponto inicial: ")) - 1
            pFinal = int(input("Ponto final: ")) - 1
            caminho = [pInicial]
            global threads
            acharCaminho(pInicial, pFinal, caminho.copy())
            for i in threads:
                i.join()
            print(
                f"De {qtde} repetições, o caminho mais rápido é {caminhoMaisRapido} com valor {menorValor}\n"
            )

        elif escolha == 6 and isinstance(matriz, list):
            matriz = 0
            print("Matriz apagada")

        else:
            if not isinstance(matriz, list):
                print("Inicialize a matriz")
            else:
                print("Escolha uma opção válida")


if __name__ == "__main__":
    main()

File: wilson_grafo.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <thread>
#include <mutex>

struct ThreadData
{
    int pInicial;
    int pFinal;
    std::vector<int> caminho;
    int valorAresta;
};

int qtde = 0;
int menorValor = INT_MAX;
std::string caminhoMaisRapido;
std::vector<std::vector<int>> matriz;
std::mutex mtx;
std::vector<std::thread> threads;

void acharCaminho(ThreadData *data)
{
    int pInicial = data->pInicial;
    int pFinal = data->pFinal;
    std::vector<int> caminho = data->caminho;
    int valorAresta = data->valorAresta;
    delete data;

    if (pInicial == pFinal)
    {
        caminho.push_back(pInicial);
        std::string caminhoStr = std::to_string(caminho[0] + 1);
        for (size_t i = 1; i < caminho.size(); i++)
        {
            caminhoStr += " -> " + std::to_string(caminho[i] + 1);
        }

        std::lock_guard<std::mutex> lock(mtx); 
        if (menorValor > valorAresta)
        {
            menorValor = valorAresta;
            caminhoMaisRapido = caminhoStr;
        }
        qtde++;
        std::cout << caminhoStr << ": " << valorAresta << "\n";
        return;
    }

    for (int i = 0; i < matriz.size(); i++)
    {
        int valor = matriz[pInicial][i];
        if (std::find(caminho.begin(), caminho.end(), i) == caminho.end() && valor > 0)
        {
            ThreadData *newData = new ThreadData{i, pFinal, caminho, valorAresta + valor};
            newData->caminho.push_back(pInicial);
            std::thread(acharCaminho, newData).detach();
        }
    }
}

std::vector<std::vector<int>> iniciarMatriz()
{
    int numRange;
    std::cout << "range: ";
    std::cin >> numRange;
    std::vector<std::vector<int>> lista(numRange, std::vector<int>(numRange, 0));
    return lista;
}

int main()
{
    while (true)
    {
        int escolha;
        std::cout << "\n1: Iniciar a Matriz\n2: Mostrar a Matriz\n3: Editar a Matriz\n4: Números Aleatórios\n5: Achar caminho\n6: Excluir a Matriz\n";
        std::cin >> escolha;

        if (escolha == 1)
        {
            matriz = iniciarMatriz();
        }
        else if (escolha == 2 && !matriz.empty())
        {
            for (const auto &row : matriz)
            {
                for (int value : row)
                {
                    std::cout << value << " ";
                }
                std::cout << "\n";
            }
        }
        else if (escolha == 3 && !matriz.empty())
        {
            char orientadoBool;
            std::cout << "É orientado ou não? (Y/N) ";
            std::cin >> orientadoBool;
            int arestas;
            std::cout << "Quantas arestas?\n";
            std::cin >> arestas;
            for (int i = 0; i < arestas; i++)
            {
                int x, y, valor;
                std::cout << "Índices e valor: ";
                std::cin >> x >> y >> valor;
                x--;
                y--;
                if (x == y)
                {
                    std::cout << "Os índice não podem ser iguais\n";
                }
                else
                {
                    matriz[x][y] = valor;
                    if (orientadoBool != 'Y')
                    {
                        matriz[y][x] = valor;
                    }
                }
            }
        }
        else if (escolha == 4 && !matriz.empty())
        {
            int lenght = matriz.size();
            for (int i = 0; i < lenght; i++)
            {
                for (int j = 0; j < lenght; j++)
                {
                    if (i != j)
                    {
                        matriz[j][i] = rand() % 10;
                    }
                }
            }
        }
        else if (escolha == 5 && !matriz.empty())
        {
            int pInicial, pFinal;
            std::cout << "Ponto inicial: ";
            std::cin >> pInicial;
            std::cout << "Ponto final: ";
            std::cin >> pFinal;
            pInicial--;
            pFinal--;
            std::vector<int> caminho = {pInicial};

            ThreadData *data = new ThreadData{pInicial, pFinal, caminho, 0};
            acharCaminho(data);

            std::this_thread::sleep_for(std::chrono::seconds(1));

            std::cout << "De " << qtde << " repetições, o caminho mais rápido é " << caminhoMaisRapido << " com valor " << menorValor << "\n";
        }
        else if (escolha == 6 && !matriz.empty())
        {
            matriz.clear();
            std::cout << "Matriz apagada\n";
        }
        else
        {
            if (matriz.empty())
            {
                std::cout << "Inicialize a matriz\n";
            }
            else
            {
                std::cout << "Escolha uma opção válida\n";
            }
        }
    }

    return 0;
}
